tool.c

```diff
@@ -221,9 +222,105 @@ compdb(int argc, char *argv[])
 	return 0;
 }
 
+static void
+targets_rule(int argc, char *argv[])
+{
+	struct edge *e = NULL;
+	for (e = alledges; e; e = e->allnext) {
+		if (argc == 2) {
+			if (e->nin == 0)
+				continue;
+
+			if (!e->in[0]->gen)
+				printf("%s\n", e->in[0]->path->s);
+		}
+		else if (argc == 3) {
+			if (e->nout == 0)
+				continue;
+
+			if (strcmp(e->rule->name, argv[2]) == 0)
+				printf("%s\n", e->out[0]->path->s);
+		}
+	}
+}
+
+static void
+targets_depth(struct node **nodes, size_t nnodes, size_t depth, size_t indent)
+{
+	for (size_t i = 0; i < nnodes; ++i) {
+		if (nodes[i]->gen == NULL)
+			continue;
+
+		for (size_t j = 0; j < indent; ++j)
+			printf("  ");
+
+		printf("%s: %s\n", nodes[i]->path->s, nodes[i]->gen->rule->name);
+
+		if (depth > 1 || depth <= 0)
+			targets_depth(nodes[i]->gen->in, nodes[i]->gen->nin,
+					depth - 1, indent + 1);
+	}
+}
+
+static void
+targets_all(void)
+{
+	struct edge *e = NULL;
+	for (e = alledges; e; e = e->allnext) {
+		if (e->nout == 0)
+			continue;
+
+		printf("%s: %s\n", e->out[0]->path->s, e->rule->name);
+	}
+}
+
+static int
+targets(int argc, char *argv[])
+{
+	size_t depth = 1;
+	if (argc >= 2) {
+		const char *mode = argv[1];
+		if (strcmp(mode, "rule") == 0) {
+			targets_rule(argc, argv);
+			return 0;
+		} else if (strcmp(mode, "depth") == 0) {
+			if (argc > 1)
+				sscanf(argv[2], "%zu", &depth);
```

Comment:
Use `strtol` here so that we can detect when we are passed an invalid depth. See the argument handling in samu.c for examples.

Reply:
Done